### PR TITLE
[bootstrap] Fix building setuptools

### LINKED ISSUE #2428: Bootstrap fails on CentOS 6
I am trying to install Easybuild on CentOS 6 and can not get it to work. I tried first on our cluster and after multiple unsuccesful attempts I installed a basic CentOS 6 VM to see if maybe we had some customizations that were incompatible, but I'm seeing the same error. Should this work or do we need to upgrade to CentOS 7?

 Here are my systems stats:
```
[mrobbert@centos6 ~]$ python -V
Python 2.6.6
[mrobbert@centos6 ~]$ type module
module is a function
module ()
{
    eval $($LMOD_CMD bash "$@") && eval $(${LMOD_SETTARG_CMD:-:} -s sh)
}
[mrobbert@centos6 ~]$ type -f module
-bash: type: module: not found
[mrobbert@centos6 ~]$ module --version

Modules based on Lua: Version 7.7.15  2018-02-13 14:12 -07:00
    by Robert McLay mclay@tacc.utexas.edu
```

```
[mrobbert@centos6 ~]$ module av EasyBuild
No modules found!
Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
```

```
[mrobbert@centos6 ~]$ which -a eb
/usr/bin/which: no eb in (/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/home/mrobbert/bin)
[mrobbert@centos6 ~]$ eb --version
-bash: eb: command not found
[mrobbert@centos6 ~]$
```

Here is the full output of the bootstrap:

```
[mrobbert@centos6 ~]$ python bootstrap_eb.py /sw
[[INFO]] EasyBuild bootstrap script (version 20180201.01, MD5: 5211e214b622e6d80614c3127eec2396)
[[INFO]] Found Python 2.6.6 (r266:84292, Aug 18 2016, 15:13:37) ; [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]

[[INFO]] Installation prefix /sw
[[INFO]] Found module command '/sw/apps/lmod/lmod/libexec/lmod' via $LMOD_CMD (Lmod), so using it.
[[INFO]] No suitable setuptools installation found, proceeding with stage 0...


[[INFO]] +++ STAGE 0: installing distribute via included (patched) distribute_setup.py...

Downloading https://easybuilders.github.io/easybuild/files/distribute-0.6.49-patched1.tar.gz
Extracting in /tmp/tmpgeoF_c
Now working in /tmp/tmpgeoF_c/distribute-0.6.49
Installing Distribute
[[INFO]] Installed setuptools version 0.6 (/tmp/tmpIbOEwo/lib/python2.6/site-packages/distribute-0.6.49-py2.6.egg/setuptools/__init__.pyc)


[[INFO]] +++ STAGE 1: installing EasyBuild in temporary dir with easy_install...

[[INFO]] installing EasyBuild with 'easy_install --quiet --upgrade --prefix=/tmp/tmpIbOEwo/eb_stage1 easybuild'
[[ERROR]] Running 'easy_install --quiet --upgrade --prefix=/tmp/tmpIbOEwo/eb_stage1 easybuild' failed: error: Could not find suitable distribution for Requirement.parse('easybuild')
Traceback (most recent call last):
  File "bootstrap_eb.py", line 359, in run_easy_install
    easy_install.main(args)
  File "/tmp/tmpIbOEwo/lib/python2.6/site-packages/distribute-0.6.49-py2.6.egg/setuptools/command/easy_install.py", line 1973, in main
    with_ei_usage(lambda:
  File "/tmp/tmpIbOEwo/lib/python2.6/site-packages/distribute-0.6.49-py2.6.egg/setuptools/command/easy_install.py", line 1954, in with_ei_usage
    return f()
  File "/tmp/tmpIbOEwo/lib/python2.6/site-packages/distribute-0.6.49-py2.6.egg/setuptools/command/easy_install.py", line 1977, in <lambda>
    distclass=DistributionWithoutHelpCommands, **kw
  File "/usr/lib64/python2.6/distutils/core.py", line 169, in setup
    raise SystemExit, "error: " + str(msg)
SystemExit: error: Could not find suitable distribution for Requirement.parse('easybuild')
```

### LINKED ISSUE #3200: bootstrap_eb.py, version 20200203.01, error installing distribute_setup.py
Trying the new `bootstrap_eb.py`, and I get this
```[bennet@lh-build tmp]$ python36 bootstrap_eb.py /sw/arcts/centos7/easybuild
[[INFO]] EasyBuild bootstrap script (version 20200203.01, MD5: a2d71121de3c3f328f4d59f9eaf27f34)
[[INFO]] Found Python 3.6.6 (default, Mar 29 2019, 00:03:27) ; [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
[[INFO]] Installation prefix /sw/arcts/centos7/easybuild
[[INFO]] Found module command '/sw/arcts/centos7/lmod/lmod/libexec/lmod' via $LMOD_CMD (Lmod), so using it.
[[INFO]] No suitable setuptools installation found, proceeding with stage 0...
[[INFO]] +++ STAGE 0: installing distribute via included (patched) distribute_setup.py...
Traceback (most recent call last):
  File "bootstrap_eb.py", line 1157, in <module>
    main()
  File "bootstrap_eb.py", line 942, in main
    distribute_egg_dir = stage0(tmpdir)
  File "bootstrap_eb.py", line 431, in stage0
    txt)
  File "/usr/lib64/python3.6/re.py", line 191, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: cannot use a string pattern on a bytes-like object
```
just downloaded from
https://raw.githubusercontent.com/easybuilders/easybuild-framework/develop/easybuild/scripts/bootstrap_eb.py

The code that it refers to, File "bootstrap_eb.py", line 431, in stage0, is
```
        txt = re.sub(r'([^\n]*)(return subprocess.call\(args)(\) == 0)',
                     r"\1f = open(os.devnull, 'w'); \2, stdout=f, stderr=f\3",
                     txt)
```
where
```
txt = DISTRIBUTE_SETUP_PY
and
if IS_PY3:
    DISTRIBUTE_SETUP_PY = DISTRIBUTE_SETUP_PY.encode('ascii')
DISTRIBUTE_SETUP_PY = codecs.decode(codecs.decode(DISTRIBUTE_SETUP_PY, "base64"), "zlib")
```
It looks like the decoding of `DISTRIBUTE_SETUP_PY` may now result in a byte-stream instead of a string, so the regular expression is carping.
```
[bennet@lh-build tmp]$ python bootstrap_eb.py /tmp/hinky
DISTRIBUTE_SETUP_PY is
<type 'str'>

[bennet@lh-build tmp]$ python3 bootstrap_eb.py /tmp/hinky
DISTRIBUTE_SETUP_PY is
<class 'bytes'>
```
Converting it to a string with `DISTRIBUTE_SETUP_PY = str(DISTRIBUTE_SETUP_PY)` results in  
```
[[INFO]] +++ STAGE 0: installing distribute via included (patched) distribute_setup.py...

Traceback (most recent call last):
  File "bootstrap_eb.py", line 1161, in <module>
    main()
  File "bootstrap_eb.py", line 942, in main
    distribute_egg_dir = stage0(tmpdir)
  File "bootstrap_eb.py", line 445, in stage0
    import distribute_setup
  File "/tmp/tmpu3v2qci7/distribute_setup.py", line 1
    b'#!python\n"""Bootstrap distribute installation\n\nIf you want to use setuptools in your package\'s setup.py, just include this\nfile in the same directory with it, and add this to the top of your setup.py::\n\n . . . .
[rest of the contents]

SyntaxError: invalid syntax
```

### PR DESCRIPTION
**Pull request by Flamefire**
Fixes #3200
Fixes #2428

This started as a fix for #3200 but falls a bit short of solving the installation for a Python3 environment without setuptools already installed.

What works/got fixed:
- Installation on Python2
  - Use https for PyPi (they dropped support for http) to fix installing/using distribute
- Use correct string type for setup_distribute unpacked blob (#3200)
- Reduce some noise when checking for setuptools
- Use system toolchain instead of dummy toolchain for EB >= 4.0

Problems to be resolved in follow-up:   
TLDR: Don't use easy_install anymore

The `distribute` package used installs an incredibly ancient version of setuptools which doesn't work out of the box anymore (see above and the need for patching setuptools and the setup_distribute script). There was/is even an ez_setup.py script which seems to superseed this: https://bootstrap.pypa.io/ez_setup.py   
However even that is deprecated: https://github.com/pypa/setuptools/issues/581

I started to use the https://bootstrap.pypa.io/get-pip.py to install pip (if missing) and then use `pip` to install setuptools just as `distribute` would have. But I hit another, currently hidden problem: The EB `bootstrap` script has a function `prep` which sets up PYTHONPATH for some temporary location where `stage{0,1,2}` are going to be installed to. It then proceeds to **import** easy_install and start installing dependencies and packages of EB (e.g. vsc-*) into that environment. The problem: This environment is **not** part of the current environment (`sys.path`). Hence easy_install fails to find the dependencies it just installed and tries to redownload them, which then fails because it doesn't find the `wheels` package which is also part of the stage0 environment (and would be part of a `system()` invoked environment) This is (at least partially) due to the mismatch of the environments, i.e. python core libs check for support of features and then run `system()` but those features are not part of that sub-process to to mismatch of PYTHONPATH vs sys.path.

IMO the easiest solution would be just use pip run via `system` and install pip if it is missing. Potentially (very probably required) also upgrade pip and setuptools into the stage1 directory.

Summary of proposed EB bootstrap procedure:
- Prepare stage1 environment in new temporary folder
  - Check for pip and if non-existant install into stage1 via get-pip script
  - Upgrade pip and setuptools with `--prefix` set to the temporary folder using pip
  - Install EB and its dependencies into that stage1 folder using pip
- In stage2
  - Use `system("eb ...")` to install the real EasyBuild via stage1 EB
  - double-check if the workarounds for pre 4.0 EB setting PYTHONPATH and force-installing vsc-install and setuptools are still required for this setup. I'd expect EB to handle that. If that is required use `pip` instead to do that.

In addition: Setup CI for the boostrap script   
- Use a dedicated .github file for those tests instead of duplicates in travis and github testing in non-clean environments
- For all python versions test in default environment AND in stripped environment where all packages are removed

### CODE REVIEW

**Comment by Flamefire** [comment]
As the bootstrap script gets deprecated closing this
